node/lib/util/synthetic_gc_util.js

```diff
@@ -0,0 +1,330 @@
+
+const assert  = require("chai").assert;
+const co = require("co");
+const NodeGit = require("nodegit");
+const SubmoduleUtil = require("./submodule_util");
+const SubmoduleFetcher = require("./submodule_fetcher");
+
+const SYNTHETIC_BRANCH_BASE = "refs/commits/";
+
+/**
+ * This class provides a way to list/delete synthetic refs, by following two
+ * main algorithms.
+ *  1) Removing refs that have persistent descendant.
+ *  2) Removing refs that are older than specific threshold date and not part
+ *     of the persistent descendant list.
+ *
+ * @SEE_ALSO: lib/synthetic_gc
+ */
+class SyntheticGcUtil {
+
+    /**
+    * Create a 'SyntheticGcUtil' object for synthetic refs manipulation.
+    */
+    constructor() {
+        this.d_visited = {};
+        this.d_simulation = true;
+    }
+
+    /**
+    * If simulation set to true, no synthetic refs are
+    * actually being removed.
+    * @param {Boolean}
+    */
+    get simulation() {
+        return this.d_simulation;
+    }
+
+    set simulation(value) {
+        this.d_simulation = value;
+    }
+
+    /**
+    * @param {Set} visited commits
+    */
+    get visited() {
+        return this.d_visited;
+    }
+
+    set visited(value) {
+        this.d_visited = value;
+    }
+
+} // SyntheticGcUtil
+
+/**
+ * Return bare submodule repository corresponding to a specified `refHeadCommit`
+ * within specified meta `repo`.
+ *
+ * @param {NodeGit.Repository}   repo
+ * @param {NodeGit.Commit}       commit
+ * @return {Number} 0 or an error code
+ */
+SyntheticGcUtil.prototype.getBareSubmoduleRepo = co.wrap(
+    function*(repo, subName, refHeadCommit) {
+
+        assert.instanceOf(repo, NodeGit.Repository);
+        assert.instanceOf(refHeadCommit, NodeGit.Commit);
+
+        const fetcher = new SubmoduleFetcher(repo, refHeadCommit);
+
+        const subUrl = yield fetcher.getSubmoduleUrl(subName);
+        const subRepo = yield NodeGit.Repository.open(subUrl);
+
+        return subRepo;
+});
+
+/**
+ * Remove synthetic ref corresponding to specified `commit` in the specified
+ * `repo`.
+ *
+ * @param {NodeGit.Repository}   repo
+ * @param {NodeGit.Commit}       commit
+ * @return {Number} 0 or an error code
+ */
+SyntheticGcUtil.prototype.removeSyntheticRef = co.wrap(
+    function(repo, commit) {
+
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(commit, NodeGit.Commit);
+
+    const synRefPath = SYNTHETIC_BRANCH_BASE + commit;
+
+    if (this.d_simulation) {
+        console.log("Removing ref: " + synRefPath);
+        return;
+    }
+    return NodeGit.Reference.remove(repo, synRefPath);
+});
+
+/**
+ * Go through the parents of `commit` of the specified `repo` and remove
+ * synthetic reference recursively if they satisfy `isDeletable` and not part of
+ * `existingReferences`.
+ *
+ * @param {NodeGit.Repository}   repo
+ * @param {NodeGit.Commit}       commit
+ * @param {Function}             isDeletable
+ * @param {String[]}             existingReferences
```

Comment:
Is existingReferences really an array of strings, and if so do you really want to do a linear search of it on line 165?